File: client/test/test-changes.js
import chai from "chai";
const expect = chai.expect;
import fs from "fs";

import { applyChange, baseChangeFunctions, consolidateChanges } from "../src/changes.js";


import { fileURLToPath } from 'url'
import { dirname, join } from 'path'
const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)


describe("Changes Tests", () => {

  const test_data_path = join(dirname(dirname(__dirname)), "common/test-data/apply-changes-test-data.json");
  const test_data = JSON.parse(fs.readFileSync(test_data_path, "utf8"));
  const input_data = test_data["input_data"];
  const tests = test_data["tests"];


  for (let i = 0; i < tests.length; i++) {
    const test = tests[i];
    const test_name = test["test_name"];
    const input_data_name = test["input_data_name"];
    const expected = test["expected_data"];

    const subject = copyObject(input_data[input_data_name]);
    it(`Apply Changes test #${i} -- ${test_name}`, () => {
      applyChange(subject, test["change"], baseChangeFunctions);
      expect(subject).to.deep.equal(expected);
    });
  }

});


function copyObject(obj) {
  return JSON.parse(JSON.stringify(obj));
}

File: server/test/test_changes.py
from copy import deepcopy
import json
import pathlib
import pytest
from fontra.core.changes import applyChange, baseChangeFunctions


testDataPath = (
    pathlib.Path(__file__).parent.parent.parent
    / "common"
    / "test-data"
    / "apply-change-test-data.json"
)

applyChangeTestData = json.loads(testDataPath.read_text(encoding="utf-8"))
applyChangeTestInputData = applyChangeTestData["inputData"]


applyChangeTestData = [
    (
        testCase["testName"],
        testCase["inputDataName"],
        testCase["change"],
        testCase["expectedData"],
    )
    for testCase in applyChangeTestData["tests"]
]


@pytest.mark.parametrize(
    "testName, inputDataName, change, expectedData", applyChangeTestData
)
def test_applyChange(testName, inputDataName, change, expectedData):
    subject = deepcopy(applyChangeTestInputData[inputDataName])
    applyChange(subject, change, baseChangeFunctions)
    assert subject == expectedData
